sympy/solvers/solveset.py

```diff
@@ -978,6 +985,331 @@ def _solveset(f, symbol, domain, _check=False):
+    These functions return either True or False.
+
+    Solving helpers: These are the helpers that solve the equation.
+    They are invoked once the equation is identified by the `identifying
+    helpers`. These functions return either the complete solution or
+    a modified version of the equation that `solveset` might better
+    handle.
+
+    * Philosophy behind the module
+
+    `\_transolve` comes into action when `solveset` is unable to solve
+    the equation as a last resort to get the solutions. The idea is to
+    get the equation in a case (Add, Pow etc.) that satisfies its
+    general form, for example exponential equations take the form as
+    `a*f(x) + b*g(x)`, where `f(x)` and `g(x)` are power terms,
+    therefore they are included in `Add` case.
+
+    Once done different `identifying helper` conditions are included
+    to check which class the equation belongs. `Solving helper` is
+    invoked once the class of the equation is identified. If the
+    equation is solved the result is returned otherwise a
+    `ConditionSet` is returned.
+
+    How `\_transolve` is better than `\_tsolve`
+    ===========================================
+
+    1) Better output
+
+    `\_transolve` provides a better output to some equations than
+    `\_tsolve`. Though result from both the functions are correct
+    it's just that the one from `\_transolve` is easy to understand and
+    appropriate. This is because `\_transolve` simplifies the result
+    wherever necessary.
+
+    Consider a simple exponential equation
+
+    >>> f = 3**(2*x) - 2**(x + 3)
+    >>> pprint(transolve(f, x, S.Reals), use_unicode=False)
+        -3*log(2)
+    {------------------}
+     -2*log(3) + log(2)
+    >>> pprint(tsolve(f, x), use_unicode=False)
+         /   3     \
+         | --------|
+         | log(2/9)|
+    [-log\2         /]
+
+    `\_transolve's` output seems more simple, readable and appropriate
+    as compared to that of `\_tsolve`.
+
+    2) Extensible
+
+    The API of `\_transolve` is designed such that it is easily
+    extensible, i.e, it is easy to add a new class of equation solver
+    without having to mess with the API or solving code of other
+    helpers. The idea behind `\_transolve` is to make seperate helpers
+    for different equation solvers. So to solve a given class of
+    equation, one helper identifies the equation and another solves
+    the equation or modifies it into a form that can be solved.
+
+    3) Modular
+
+    `\_transolve` is designed to be modular i.e, for every class of
+    equation a separate helper for identification and solving is
+    implemented. This makes it easy to change or modify any of the
+    method implemented directly in the helpers without interfering
+    with the actual structure of the API.
+
+    4) Less complex API
+
+    The API of `\_transolve` is designed in a way that it is less
+    complex to read and understand unlike in `\_tsolve` where the
```

Comment:
It sounds like you are talking about the code rather than the API. I really doubt transolve is going to solve an equation with less recursion than tsolve does. It's just that any modification steps are located outside of transolve, isn't it?

Reply:
> I really doubt transolve is going to solve an equation with less recursion than tsolve does

Yes this could be, still `transolve` solves equation faster than `_tsolve`

```python
>>> import time
>>> f = 3**(2*x) - 2**(x + 3)

>>> start_time = time.time()
>>> _tsolve(f, x)
>>> print("--- %s seconds ---" % (time.time() - start_time))
--- 0.4206206798553467 seconds ---
>>> start_time = time.time()
>>> _transolve(f, x, S.Reals)
>>> print("--- %s seconds ---" % (time.time() - start_time))
--- 0.05263113975524902 seconds ---
```
One reason is probably a lot of function calls.

> It sounds like you are talking about the code rather than the API

I am in the impression that when we talk about API, we are indirectly talking about the code, because code is what API is built with. Am I getting this wrong?

> It's just that any modification steps are located outside of transolve, isn't it?

Yes.

Reply:
> Am I getting this wrong?

In the context of SymPy I would think of the API as the information (list of arguments) that are required for calling a function. What the function does between receiving the request and returnin a result is not the API (as I, not a profefssional in this field, have come to understand the term). So take that fo what it's worth.

Reply:
I guess we should remove this point.